Python/Product/PythonTools/PythonTools/Intellisense/AnalysisEntry.cs

```diff
@@ -83,32 +180,38 @@ internal sealed class AnalysisEntry {
             return AnalysisCookie?.GetLine(line);
         }
 
-        internal void OnParseComplete() {
-            var bufferParser = BufferParser;
-            if (bufferParser != null) {
-                foreach (var buffer in bufferParser.Buffers) {
-                    var events = buffer.GetParseTreeRegistrations();
-                    foreach (var notify in events) {
-                        notify(this);
-                    }
+        internal async void OnParseComplete() {
+            BufferParser bufferParser;
+            try {
+                bufferParser = await GetBufferParserAsync();
+            } catch (OperationCanceledException) {
+                return;
+            }
+            foreach (var buffer in bufferParser.Buffers) {
+                var events = buffer.GetParseTreeRegistrations();
+                foreach (var notify in events) {
+                    notify(this);
                 }
             }
         }
 
-        internal void OnNewAnalysisEntry() {
-            var bufferParser = BufferParser;
-            if (bufferParser != null) {
-                foreach (var buffer in bufferParser.Buffers) {
-                    var events = buffer.GetNewAnalysisEntryRegistrations();
-                    foreach (var notify in events) {
-                        notify(this);
-                    }
+        internal async void OnNewAnalysisEntry() {
+            BufferParser bufferParser;
+            try {
+                bufferParser = await GetBufferParserAsync();
+            } catch (OperationCanceledException) {
+                return;
+            }
+            foreach (var buffer in bufferParser.Buffers) {
+                var events = buffer.GetNewAnalysisEntryRegistrations();
+                foreach (var notify in events) {
+                    notify(this);
                 }
             }
         }
 
         public int GetBufferId(ITextBuffer buffer) {
-            var bufferParser = BufferParser;
+            var bufferParser = _bufferParser;
```

Comment:
>_bufferParser [](start = 31, length = 13)

this here and in the function below are the only places accessing _bufferParser without lock (_bufferParserLock)